src/cascadia/WindowsTerminal/WindowEmperor.cpp

```diff
@@ -0,0 +1,690 @@
+
+WindowEmperor::WindowEmperor() noexcept :
+    _app{}
+{
+    _manager.FindTargetWindowRequested([this](const winrt::Windows::Foundation::IInspectable& /*sender*/,
+                                              const winrt::Microsoft::Terminal::Remoting::FindTargetWindowArgs& findWindowArgs) {
+        {
+            const auto targetWindow = _app.Logic().FindTargetWindow(findWindowArgs.Args().Commandline());
+            findWindowArgs.ResultTargetWindow(targetWindow.WindowId());
+            findWindowArgs.ResultTargetWindowName(targetWindow.WindowName());
+        }
+    });
+
+    _dispatcher = winrt::Windows::System::DispatcherQueue::GetForCurrentThread();
+}
+
+WindowEmperor::~WindowEmperor()
+{
+    _app.Close();
+    _app = nullptr;
+}
+
+void _buildArgsFromCommandline(std::vector<winrt::hstring>& args)
+{
+    if (auto commandline{ GetCommandLineW() })
+    {
+        auto argc = 0;
+
+        // Get the argv, and turn them into a hstring array to pass to the app.
+        wil::unique_any<LPWSTR*, decltype(&::LocalFree), ::LocalFree> argv{ CommandLineToArgvW(commandline, &argc) };
+        if (argv)
+        {
+            for (auto& elem : wil::make_range(argv.get(), argc))
+            {
+                args.emplace_back(elem);
+            }
+        }
+    }
+    if (args.empty())
+    {
+        args.emplace_back(L"wt.exe");
+    }
+}
+
+bool WindowEmperor::HandleCommandlineArgs()
+{
+    std::vector<winrt::hstring> args;
+    _buildArgsFromCommandline(args);
+    auto cwd{ wil::GetCurrentDirectoryW<std::wstring>() };
+
+    Remoting::CommandlineArgs eventArgs{ { args }, { cwd } };
+
+    const auto isolatedMode{ _app.Logic().IsolatedMode() };
+
+    const auto result = _manager.ProposeCommandline(eventArgs, isolatedMode);
+
+    if (result.ShouldCreateWindow())
+    {
+        CreateNewWindowThread(Remoting::WindowRequestedArgs{ result, eventArgs }, true);
+
+        _manager.RequestNewWindow([this](auto&&, const Remoting::WindowRequestedArgs& args) {
+            CreateNewWindowThread(args, false);
+        });
+
+        _becomeMonarch();
+    }
+    else
+    {
+        const auto res = _app.Logic().GetParseCommandlineMessage(eventArgs.Commandline());
+        if (!res.Message.empty())
+        {
+            AppHost::s_DisplayMessageBox(res);
+            ExitThread(res.ExitCode);
+        }
+    }
+
+    return result.ShouldCreateWindow();
+}
+
+void WindowEmperor::WaitForWindows()
+{
+    MSG message;
+    while (GetMessage(&message, nullptr, 0, 0))
+    {
+        TranslateMessage(&message);
+        DispatchMessage(&message);
+    }
+}
+
+void WindowEmperor::CreateNewWindowThread(Remoting::WindowRequestedArgs args, const bool /*firstWindow*/)
+{
+    Remoting::Peasant peasant{ _manager.CreatePeasant(args) };
+
+    auto window{ std::make_shared<WindowThread>(_app.Logic(), args, _manager, peasant) };
+
+    window->Started([this, sender = window]() -> winrt::fire_and_forget {

```

Comment:
Should this use `get_weak()` instead of `this`, or can you guarantee that the main thread waits for all coroutines to finish before it kills the `WindowEmperor` instance? Same below (everywhere it says `[this`).

Better yet, instead of using callbacks, you could also just pass a `get_weak()` into the constructor of `WindowThread` which then stores it and uses it to do the below itself.

Reply:
Interestingly no, we can't use a `get_weak`, because the emperor isn't a winrt class.

> can you guarantee that the main thread waits for all coroutines to finish before it kills the WindowEmperor instance

That's an interesting inquiry and I'm pretty confident the answer is no. Pretty sure the `App` teardown pumps messages for a while to let XAML finish it's business, including callbacks. These are a little different though, cause these are callbacks from the WindowThreads to the emperor, and theoretically, if the emperor is getting dtor'd, it has already released all the WindowThreads...

Reply:
But `WindowThread` and `WindowEmporer` are in the same project. Why is that a layering violation then? And why is violating layers a concern anyways? It's not a public API after all.

BTW for classes that aren't WinRT, I think we can use `std::enable_shared_from_this` liberally, since our lifetime model in the UI code is a bit unclear and this would allow us to use shared pointers like we can use them for WinRT (via `shared_from_this()` and `weak_from_this()`). The overhead of shared pointers is basically negligible compared to `std::unique_ptr` for this purpose. (I'm just bringing this up as a general idea, not in particular for this code.)

Reply:
That's basically the "dependency inversion principle" (the D in SOLID), right? I've always felt like its benefit was a little overstated, if applied rigorously. It's not like we're gonna have 2 different Emperors in the foreseeable future after all. Unless for unit testing of course... hmm. Oh well, I basically already made my point and I'll leave it up to your consideration now. 😅